core/unix/signal.c

```diff
@@ -6693,6 +6700,141 @@ execute_native_handler(dcontext_t *dcontext, int sig, sigframe_rt_t *our_frame)
     }
 }
 
+// Attempts to execute the native signal handler by reusing the current frame
+// which is assumed to be at xsp. If it succeeds, it does not return. Otherwise
+// it returns a bool: false if DR should continue trying to deliver the signal.
+static bool
+execute_native_handler_using_cur_frame(dcontext_t *dcontext, int sig, byte *xsp)
+{
+    // For a signal that arrived when the thread is native and the process is in
+    // the middle of detaching, we should try to reuse the current DR signal
+    // frame to deliver it if possible. This is to avoid using up too much app
+    // stack, which has been seen to cause a stack overflow in the app (i#6814).
+    // If we cannot optimize, we fall back to the regular
+    // execute_native_handler() on returning.
+    //
+    // Summary of relevant events when DR starts detaching:
+    // D: Detacher thread sends suspend signal to all other threads
+    // O: each of the Other threads receive the suspend signal. Note that DR's
+    //    signal handling configuration blocks all signals except SIGSEGV and
+    //    the suspend signal. So at this point Other threads cannot receive
+    //    app signals. Other threads signal to the Detacher thread that they're
+    //    suspended and awaiting wakeup call to continue detaching (sig_detach).
+    // D: Detacher thread sets doing_detach to true. A side effect of this is
+    //    that get_thread_private_dcontext() will return NULL when any Other
+    //    thread tries to get its private dcontext.
+    // D: Detacher thread wakes up each Other thread, telling it to detach.
+    // O: Each Other thread wakes up, does sig_detach which reinstates the app
+    //    signal stack (if available) and the app's blocked signal set, lets the
```

Comment:
I thought you removed the sigprocmask call

Reply:
The app's blocked signal set is reinstated to the signal frame; the following sigreturn would actually restore it. I'll clarify.

Reply:
Done